sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala

```diff
@@ -138,7 +138,8 @@ case class Percentile(
   override def update(
       buffer: OpenHashMap[Number, Long],
       input: InternalRow): OpenHashMap[Number, Long] = {
-    val key = child.eval(input).asInstanceOf[Number]
+    val scalaValue = CatalystTypeConverters.convertToScala(child.eval(input), child.dataType)
```

Comment:
I'm not 100% sure though, it seems the cost of converting `Decimal` to `BigDecimal` every-time is some higher than that of using catalyst values as it is.